community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -470,6 +472,92 @@ def update_topology(lkp: util.Lookup) -> None:
+
+
+def create_reservation(lkp: util.Lookup, reservation_name: str, node: str, start_time: datetime) -> None:
+    # Format time to be compatible with slurm reservation.
+    formatted_start_time = start_time.strftime('%Y-%m-%dT%H:%M:%S')
+    util.run(f"{lkp.scontrol} create reservation user=slurm starttime={formatted_start_time} duration=180 nodes={node} reservationname={reservation_name}")
+
+
+def get_slurm_reservation_maintenance(lkp: util.Lookup) -> Dict[str, datetime]:
+    res = util.run(f"{lkp.scontrol} show reservation --json")
+    all_reservations = json.loads(res.stdout)
+    reservation_map = {}
+
+    for reservation in all_reservations['reservations']:
+        reservation_name = reservation['name']
+        nodes = reservation['node_list']
+        time = datetime.fromtimestamp(reservation['start_time']['number'])
+
+        if reservation_name is None or nodes is None or time is None:
+          continue
+
+        # There should be 1 reservation per node per maintenance.
+        if reservation['node_count'] != 1:
+          continue
+
+        # Check if the reservation is for scheduled maintenance.
+        maintenance_reservation = nodes + "_maintenance"
+        if reservation_name != maintenance_reservation:
+          continue
+
+        reservation_map[reservation_name] = time
+
+    return reservation_map
+
+
+def get_upcoming_maintenance(lkp: util.Lookup) -> Dict[str, Tuple[str, datetime]]:
+    upc_maint_map = {}
+
+    for node, properties in lkp.instances().items():
+        if 'upcomingMaintenance' in properties:
+          time = datetime.strptime(rfc3339_time, properties['upcomingMaintenance']['startTimeWindow']['earliest'])
+          upc_maint_map[node + "_maintenance"] = (node, time)
+
+    return upc_maint_map
+
+# Sync maintenance reservation gets upcoming maintenance notification and
+# updates slurm reservation for the node during scheduled maintenance window.
+def sync_maintenance_reservation(lkp: util.Lookup) -> None:
+    upc_maint_map = get_upcoming_maintenance(lkp)  # map reservation_name -> (node_name, time)
+    log.debug(f"upcoming-maintenance-vms: {upc_maint_map}")
+
+    # Get current slurm reservation for maintenance.
+    # map reservation_name -> time
+    # reservation_name should be of format nodes_maintenance.
+    curr_reservation_map = get_slurm_reservation_maintenance(lkp)
+    log.debug(f"curr-reservation-map: {curr_reservation_map}")
+
+    # Separate reservations for which maintenance is already finished.
+    del_reservation = set(curr_reservation_map.keys() - upc_maint_map.keys())
+    create_reservation_map = {}
+
+    # Iterate upcoming maintenance and update reservation.
+    # Also remove reservation from the queue after update.
+    for res_name, (node, time) in upc_maint_map.items():
+      if res_name in curr_reservation_map:
+        diff = curr_reservation_map[res_name] - time
+        if abs(diff) <= datetime.timedelta(seconds=60):
```

Comment:
Is there a particular reason for such large delta? Would 1 second be sufficient?

Reply:
I was looking at this as error window. But since the script runs every 30 seconds, it doesn't make sense to have 60 seconds threshold. Updated it to 1 sec.